### PR TITLE
Config file refactor & command line options

### PR DESCRIPTION
**Pull request by StratusFearMe21**
The config file and the command line options are highly integrated so I included both in this PR. This PR makes the config file mimic the layout of the settings menu in the actual GUI

These are the default options
```toml
[synth]
synth = "xsynth"
buffer_ms = 10.0
sfz_path = ""
limit_layers = true
layer_count = 4
fade_out_kill = false
linear_envelope = false
use_effects = true

[synth.vel_ignore]
hi = 0
lo = 0

[midi]
note_speed = 0.25
random_colors = false
midi_loading = "ram"

[midi.key_range]
hi = 0
lo = 127

[visual]
audio_only = false
bg_color = "#1e1e1e"
bar_color = "#910000"
show_top_pannel = true
show_statistics = true
fullscreen = false
```
There are also command line options, each option containing a short and long help text. This is the short help
```
Usage: wasabi [OPTIONS] [midi-file]

Arguments:
  [midi-file]  The MIDI file to immediately begin playing

Options:
  -S, --synth <synth>                The synthesizer to use
  -b, --buffer-ms <buffer-ms>        The amount of time events are held in the buffer
  -s, --sfz-path <sfz-path>          The path to an SFZ SoundFont
      --dont-limit-layers            Do not apply the layer limit to the synth
  -l, --layer-count <layer-count>    The amount of layers that the synthesizer can create
  -v, --vel-ignore <vel-ignore>      The range of note velocities that the synth will discard
  -F, --fade-out-kill                Once a voice is killed, fade it out
  -L, --linear-envelope              Adjust the rate of decay on the synth's voices
  -N, --no-effects                   Disable the synth's effects
  -n, --note-speed <note-speed>      The speed that the notes travel on-screen
  -r, --random-colors                Make each channel a random color
  -k, --key-range <key-range>        The key range of the on-screen piano keyboard
  -m, --midi-loading <midi-loading>  How the MIDI is loaded into `wasabi`
      --bg-color <bg-color>          The window background
      --bar-color <bar-color>        The color of the bar just above the piano
      --hide-top-pannel              Hide the top panel
      --hide-statistics              Hide the statistics window
  -f, --fullscreen                   Start `wasabi` in fullscreen
  -h, --help                         Print help (see more with '--help')
  -V, --version                      Print version
```
This is the long help
```
Usage: wasabi [OPTIONS] [midi-file]

Arguments:
  [midi-file]
          This MIDI file is played immediately after the app's launch. This argument is required to use the `--audio-only` option

Options:
  -S, --synth <synth>
          The synthesizer that is used to play the MIDI. This can either be XSynth (recommended) or KDMAPI. KDMAPI only works if you have OmniMidi installed, and are using Windows

  -b, --buffer-ms <buffer-ms>
          The amount of time that events are held in the buffer before being played. Higher numbers may increase performance but also increase latency.

  -s, --sfz-path <sfz-path>
          The path to any SFZ soundfont. In audio only mode a soundfont must be passed either via the config file, or
                                  this command line option. In the GUI you can set this under `Open Synth Settings > SFZ Path`

      --dont-limit-layers
          This allows the synth to create as many layers as it needs to play the MIDI file faithfully. Only turn this on if your MIDI sounds bad, or your computer is running on GFuel

  -l, --layer-count <layer-count>
          The amount of virtual pianos that the synth can create to play your piece. A reasonable number for this value is the number of CPU cores your computer has.

  -v, --vel-ignore <vel-ignore>
          Two numbers, comma seperated, that represent a range of velocities that the synth will discard, making notes in the range inaudible.

  -F, --fade-out-kill
          Once the synthesizer kills one of it's voices, it will fade it out as opposed to simply cutting it off

  -L, --linear-envelope
          Switch the synth's voice's rate of decay from exponential to linear. This may be bring a performance improvement on some systems.

  -N, --no-effects
          Disable the effects that the synthesizer applies to the final audio render. These effects include a limiter to keep the audio from clipping, and a cutoff

  -n, --note-speed <note-speed>
          The speed at which the notes will move across the screen. This makes the notes physically longer, causing them to move faster on-screen

  -r, --random-colors
          This causes each of the note channels to become a random color

  -k, --key-range <key-range>
          Two numbers, comma seperated, that describe the range of keys to be shown on the on-screen piano keyboard, the range must be less than 255 and more than 0

  -m, --midi-loading <midi-loading>
          The method in which the MIDI file is loaded into `wasabi`, the two possible options are `ram`, which loads the MIDI file entirely into RAM before beginning playback; and `live` which will read the MIDI file as it's being played back. The latter method is for using with systems with low memory

      --bg-color <bg-color>
          A hex color string describing the background color of the window

      --bar-color <bar-color>
          A hex color string describing the color of the bar just above the on-screen piano keyboard

      --hide-top-pannel
          Hides the top panel from view when the app opens. It can be un-hidden with Ctrl+F

      --hide-statistics
          Hides the statistics window from view when the app opens. It can be un-hidden with Ctrl+G

  -f, --fullscreen
          Starts `wasabi` in fullscreen mode. `wasabi` will use borderless fullscreen mode on Linux systems running Wayland, and exclusive fullscreen mode for everyone else

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version
```

### CODE REVIEW

File: src/settings.rs [hunk 1]
```diff
@@ -1,160 +1,336 @@
...
+pub struct WasabiSettings {
+    pub synth: SynthSettings,
+    pub midi: MidiSettings,
+    pub visual: VisualSettings,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub load_midi_file: Option<String>,
 }
 
 impl Default for WasabiSettings {
     fn default() -> Self {
```
Comment (MyBlackMIDIScore) [inline src/settings.rs:280]: I think it would be better to make the `Default` implementation for each struct separately.

File: src/settings.rs [hunk 2]
```diff
@@ -1,160 +1,336 @@
...
+    #[serde(with = "range_serde")]
+    pub vel_ignore: RangeInclusive<u8>,
     pub fade_out_kill: bool,
     pub linear_envelope: bool,
     pub use_effects: bool,
-    pub vel_ignore: RangeInclusive<u8>,
-    pub synth: usize,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
```
Comment (MyBlackMIDIScore) [inline src/settings.rs:270]: You should also add `#[serde(default)]` to this and the other option structs so serde can load the default value for missing options (so the config is not reset when adding a new option).

File: src/settings.rs [hunk 3]
```diff
@@ -163,34 +339,299 @@ impl WasabiSettings {
...
+            .version(env!("CARGO_PKG_VERSION"))
+            .author(env!("CARGO_PKG_AUTHORS"))
+            .about(env!("CARGO_PKG_DESCRIPTION"))
+            .arg(
+                Arg::new("synth")
+                    .help("The synthesizer to use")
+                    .long_help(
+                        "The synthesizer that is used to play the MIDI. \
+                        This can either be XSynth (recommended) or KDMAPI. KDMAPI \
+                        only works if you have OmniMidi installed, and are using Windows",
```
Comment (MyBlackMIDIScore) [inline src/settings.rs:353]: Small detail, capitalize as `OmniMIDI`

File: src/settings.rs [hunk 4]
```diff
@@ -163,34 +339,299 @@ impl WasabiSettings {
...
+                        MIDI sounds bad, or your computer is running on GFuel",
+                    )
+                    .long("dont-limit-layers")
+                    .action(ArgAction::SetFalse),
+            )
+            .arg(
+                Arg::new("layer-count")
+                    .help("The amount of layers that the synthesizer can create")
+                    .long_help(
+                        "The amount of virtual pianos that the synth can create to \
```
Comment (MyBlackMIDIScore) [inline src/settings.rs:399]: I think it's better to write this as "The maximum amount of voices the synth is allowed to create per key per channel." or something
Also idk about "A reasonable number for this value is the number of CPU cores your computer has."

File: src/settings.rs [hunk 5]
```diff
@@ -163,34 +339,299 @@ impl WasabiSettings {
...
+                    )
+                    .short('L')
+                    .long("linear-envelope")
+                    .action(ArgAction::SetTrue),
+            )
+            .arg(
+                Arg::new("no-effects")
+                    .help("Disable the synth's effects")
+                    .long_help(
+                        "Disable the effects that the synthesizer applies to the final audio \
```
Comment (MyBlackMIDIScore) [inline src/settings.rs:445]: The limiter is not affected by this.
These effects only include those applied to individual voices (currently only high/low pass filter)

File: src/settings.rs (hunk 2 above)
Comment (StratusFearMe21) [reply src/settings.rs]: I've resolved this one by creating a migration system, the old config format will be migrated to the new one if it's detected

**Review by arduano — APPROVED** [review]

**Pull request merged**